### PR TITLE
jubna amp ads

### PR DESCRIPTION
**Pull request by anshuman91**
Please check and approve jubna amp ads.

### CODE REVIEW

**Comment by anshuman91** [comment]
Hello @jasti @lannka ,

Our AMP ad implementation has passed all required test. Please merge this in the main amphtml project.

**Comment by lannka** [comment]
@anshuman91 thank you for the integration!

could you follow the [instructions](https://github.com/ampproject/amphtml/blob/master/ads/README.md#developer-guidelines-for-a-pull-request) to implement the render-start API?

**Comment by torch2424** [comment]
@anshuman91

Thank you for your contribution! And thanks for suggesting the renderStart @lannka !

I also noticed a very large diff on: examples/ads/ads.amp.html . Did your text editor do changes on save? or perhaps you haven't pulled in the latest master onto your branch/fork?

Thanks!

File: ads/jubna.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
```
Comment (torch2424) [inline ads/jubna.js]: Incorrect Copyright year 😄

Comment (torch2424) [inline ads/jubna.js]: Incorrect Copyright year 😄

File: ads/jubna.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+<!---
+Copyright 2017 The AMP HTML Authors. All Rights Reserved.
```
Comment (torch2424) [inline ads/jubna.md]: Incorrect Copyright year 😄

File: examples/ads/ads.amp.html [hunk 3]
```diff
@@ -0,0 +1,1937 @@
+<!doctype html>
```
Comment (torch2424) [inline examples/ads/ads.amp.html]: I don't think this file should have such a large diff. Have you pulled in the latest master? Or perhaps your text editor did some automatic formatting?

Oh! This file was incorrectly added. Seems like you have both `examples/ads/ads.amp.html` and `examples/ads.amp.html`. You only need `examples/ads.amp.html`. Please remove this file 😄 👍

Comment (anshuman91) [reply examples/ads/ads.amp.html]: Hello @torch2424 ,

I have removed the file `examples/ads/ads.amp.html` from amp code. Please review the code again.

**Comment by anshuman91** [comment]
> @anshuman91
> 
> Thank you for your contribution! And thanks for suggesting the renderStart @lannka !
> 
> I also noticed a very large diff on: examples/ads/ads.amp.html . Did your text editor do changes on save? or perhaps you haven't pulled in the latest master onto your branch/fork?
> 
> Thanks!

Yes, I have pulled the latest code from your master repository. So now it shouldn't show you the large difference in any file.

File: ads/jubna.js (hunk 1 above)
Comment (anshuman91) [reply ads/jubna.js]: I have changed the copyright year.

**Comment by torch2424** [comment]
@anshuman91 Thank you for making the changes, been quite busy lately, but will take a look at this soon. Thank you! 😄

**Comment by torch2424** [comment]
@anshuman91

So I re-ran the travis a few times, and I think there is an issue with your integration tests. Could you merge in the latest master into your fork? Then we can finish off this review. 😄 

Thank you very much! And thank you for working with us 👍

**Comment by torch2424** [comment]
Hello @anshuman91 ,

Just checking back in, whenever this is ready, feel free to mention me.

Thanks again for the contribution! 😄

File: ads/_config.js [hunk 4]
```diff
@@ -534,6 +534,8 @@ export const adConfig = {
     renderStartImplemented: true,
   },
 
+  'jubna': {},
```
Comment (torch2424) [inline ads/_config.js]: If you have implemented `renderStart` by @lannka suggetsion, you will want to add `renderStartImplemented: true` here 😄

File: examples/ads.amp.html [hunk 5]
```diff
@@ -1130,6 +1130,14 @@ <h2>Innity</h2>
       data-channel="NEWS">
```
Comment (torch2424) [inline examples/ads.amp.html]: Seems like you forgot to add yourself to the example dropdown for all of the vendors 😄 

See: https://github.com/ampproject/amphtml/pull/20367/files

**Comment by anshuman91** [comment]
> So for the current issues with the tests. That is on us. The integration tests are kind of hit/miss, depending on the way it is run. So I went ahead and re-ran that, and it should be fine 😄
> 
> Left a few more comments on the PR on some things that are missing.
> 
> Also, wen't ahead and started testing things, and noticed ads weren't loading on the example page. Do you have an idea why?
> 
> screen shot 2019-01-30 at 2 52 32 pm
> 
> Thanks! 😄

Hello @torch2424 ,

I have fixed all the errors which were appearing in browser console and now everything seems working fine. Would you please review the implementation now and list our ad network in your AMP project.

**Comment by anshuman91** [comment]
screenshot 2019-02-04 at 4 01 48 pm
screenshot 2019-02-04 at 4 02 03 pm

**Comment by anshuman91** [comment]
Hello @torch2424,

Travis still showing one job failed so when I opened job log then it was completed successfully.
screenshot 2019-02-04 at 4 11 00 pm 

Thank you. :)

File: examples/ads.amp.html (hunk 5 above)
Comment (anshuman91) [reply examples/ads.amp.html]: I have added Jubna in all required pages now.

File: ads/_config.js (hunk 4 above)
Comment (anshuman91) [reply ads/_config.js]: No, I didn't use renderStart while implementation.

File: examples/ads.amp.html (hunk 5 above)
Comment (anshuman91) [reply examples/ads.amp.html]: https://github.com/ampproject/amphtml/pull/20109/files

**Comment by torch2424** [comment]
Hello!

So the travis issue was a flake on our end. My apologies for that!

And everything is looking good implementation wise!

screen shot 2019-02-04 at 12 19 15 pm

But it seems like you haven't implemented one last thing I asked for. See this line here: https://github.com/ampproject/amphtml/blob/master/examples/ads.amp.html#L153

Doesn't seem like you added yourself to the dropdown that lets you select the vendor you would like to see. Once you implement that, this can be merged. Thank you! 😄

File: examples/ads.amp.html (hunk 5 above)
Comment (torch2424) [reply examples/ads.amp.html]: P.S This is the original comment I made asking about the dropdown 😄

**Comment by anshuman91** [comment]
> Hello!
> 
> So the travis issue was a flake on our end. My apologies for that!
> 
> And everything is looking good implementation wise!
> 
> screen shot 2019-02-04 at 12 19 15 pm
> 
> But it seems like you haven't implemented one last thing I asked for. See this line here: https://github.com/ampproject/amphtml/blob/master/examples/ads.amp.html#L153
> 
> Doesn't seem like you added yourself to the dropdown that lets you select the vendor you would like to see. Once you implement that, this can be merged. Thank you! 😄

Hello @torch2424,

Now I have added the Jubna in the dropdown. Please review the code again.

Thanks.

**Comment by torch2424** [comment]
@anshuman91 Hello,

Sorry for the delay on this, it's an issue on our side. You code and implementation all look correct, but our travis system appears to be flaking on your PR. This is probably on our end, so I'll keep you updated, and merge as soon as we can 😄 

Thanks for your patience!

**Comment by anshuman91** [comment]
> @anshuman91 Hello,
> 
> Sorry for the delay on this, it's an issue on our side. You code and implementation all look correct, but our travis system appears to be flaking on your PR. This is probably on our end, so I'll keep you updated, and merge as soon as we can 😄
> 
> Thanks for your patience!

Thank you so much @torch2424. Waiting for approval. :smiley:

**Comment by torch2424** [comment]
@anshuman91 Hello! So it seems like we issued a fix for travis. but things are still acting up. Can you pull in the latest master? Thank you soooo much for your patience! 😄

**Comment by anshuman91** [comment]
> @anshuman91 Hello! So it seems like we issued a fix for travis. but things are still acting up. Can you pull in the latest master? Thank you soooo much for your patience! 😄

Hello @torch2424 ,

I have taken the latest pull from your master branch. Please check. 
Thanks

**Comment by anshuman91** [comment]
Hello @torch2424,
After taking the pull from master brach one more test start falling. Now one visual change got failed and it also doesn't belong to Jubna implementation.

**Comment by anshuman91** [comment]
Hello @torch2424,

Currently, all checks are passing and Percy check throwing one visual issue which doesn't belong to my (Jubna) implementation. 
screenshot 2019-02-11 at 6 55 57 pm

**Comment by torch2424** [comment]
@anshuman91 Ah the percy check can be solved by me no worries. Just the travis check needs to be done. Seems like it is working, so once this is done building we should be good to go!

Thank you sooo much again for your patience! 😄

**Comment by torch2424** [comment]
@rsimha So this is the PR that has been stuck flaking consistently for more than a handful of days now. I've re-run this travis a ton of times. 

I opened/closed the PR, and got the weird travis error fixed, but it's consistently flaking on like 1-2 integration tests every run.

Is there anything you can do to help get this in? I just re-ran at the time of this posting.

Thanks! 😄

**Comment by rsimha** [comment]
Seems like tests passed?

**Comment by anshuman91** [comment]
> @anshuman91 Ah the percy check can be solved by me no worries. Just the travis check needs to be done. Seems like it is working, so once this is done building we should be good to go!
> 
> Thank you sooo much again for your patience! 😄
> 
> EDIT: Gonna play around with this PR (close/open, add some commits, etc..) to approve this.

Thank you so much @torch2424 . Now all check start passing. Please check and approved the pull reequest. Our many clients are waiting for Jubna AMP code. 

Thanks

**Comment by anshuman91** [comment]
> Seems like tests passed?

Hello @rsimha,

Thank you so much. Yes, now all tests are passed.

**Comment by torch2424** [comment]
@rsimha Of course it passes once I mentioned you 😂 

Tests must have gotten scared once they knew you were coming 😂 

@anshuman91 Thank you sooooooo much for your patience! Merging this in! 😄 🎉 

Oh wait it seems you merged the ltest master, will merge once the travis ends there 😂 I have it from here no worries on updating master going forward 👍

**Comment by anshuman91** [comment]
> @rsimha Of course it passes once I mentioned you 😂
> 
> Tests must have gotten scared once they knew you were coming 😂
> 
> @anshuman91 Thank you sooooooo much for your patience! Merging this in! 😄 🎉
> 
> Oh wait it seems you merged the ltest master, will merge once the travis ends there 😂 I have it from here no worries on updating master going forward 👍

Hello @torch2424 ,

All checks have passed. Please merge my implementation. 
Thanks.

**Comment by anshuman91** [comment]
Hello @torch2424, @jasti,  @lannka, 

Thank you so much for merging my pull request. 

After the approval is there any other process which I need to complete first. 

Because when I used Jubna AMP code eg.
`<amp-embed` width="200" height="200" 
type="jubna"
layout="responsive"
data-wid="12899"
data-pid="2781">
</amp-embed>`

This code throwing error in the browser console `Unknown ad type "jubna"​​`

Here I am adding all the error's screenshot
screenshot 2019-02-13 at 6 44 15 pm
screenshot 2019-02-13 at 6 43 53 pm
screenshot 2019-02-13 at 5 54 02 pm
screenshot 2019-02-13 at 5 53 53 pm
screenshot 2019-02-13 at 5 53 44 pm




. Please check this issue and let me know where I am doing wrong.

**Comment by anshuman91** [comment]
Here I am adding few sites where our AMP code is live so you can see issue directly. 
http://dev01.eremnews.com/test-amp.html
http://dev02.foochia.com/test-amp.html
https://goo.gl/AyXHfg

They could remove the codes of Jubna AMP from their site because it is not working currently so it is possible not to find the code when you are checking these sites. So I added the screenshots of all errors which occurring while running Jubna AMP code.

**Comment by torch2424** [comment]
@anshuman91

Hello,

So this is happening because we do releases on a 2 week cycle. Where the first week, is our Dev / Canary channel, and the second week is when it will go out to production. 

So you should be able to use your amp-ad type in production in about 2 weeks.

Thanks again for your contribution! 😄

**Pull request merged**